### PR TITLE
feat: 2fa management on settings page

### PR DESCRIPTION
**Pull request by zlwaterfield**
## Problem

Our 2FA experience is not good, we use a pre generated server side UI that's buggy and doesn't let users view, generated or rolled backup codes or disable 2FA

## Changes
This PR updates all the 2FA functionality to be on the PostHog user settings page (view backup codes, generate backup codes, disable 2FA)

Before:
Screenshot 2024-11-28 at 5 28 03 PM
Screenshot 2024-11-28 at 5 52 36 PM

After:
Screenshot 2024-11-28 at 5 28 14 PM
Screenshot 2024-11-28 at 5 29 04 PM
Screenshot 2024-11-28 at 5 27 11 PM
Screenshot 2024-11-28 at 5 30 36 PM
Screenshot 2024-11-28 at 5 30 32 PM
Screenshot 2024-11-28 at 5 27 51 PM

## Does this work well for both Cloud and self-hosted?

It doesn't have an impact.

## How did you test this code?

Manually and behind ff.

### CODE REVIEW

File: posthog/api/authentication.py [hunk 1]
```diff
@@ -247,16 +248,22 @@ def create(self, request: Request, *args: Any, **kwargs: Any) -> Any:
...
+            totp_device = default_device(user)
+            if totp_device:
+                is_allowed = totp_device.verify_is_allowed()
+                if not is_allowed[0]:
+                    raise serializers.ValidationError(detail="Too many attempts.", code="2fa_too_many_attempts")
+                if totp_device.verify_token(request.data["token"]):
+                    return self._token_is_valid(request, user, totp_device)
+                totp_device.throttle_increment()
+
+            # Then try backup codes
```
Comment (zlwaterfield) [inline posthog/api/authentication.py:261]: Need to write tests for this change

**Comment by raquelmsmith** [comment]
Amazing

File: frontend/src/scenes/settings/user/TwoFactorAuthentication.tsx [hunk 2]
```diff
@@ -1,53 +1,145 @@
...
+                            you lose access to your authentication device.
+                        </p>
+                        {status?.backup_codes?.length ? (
+                            <div className="bg-bg-3000 p-4 rounded font-mono space-y-1 relative">
+                                <LemonButton
+                                    icon={<IconCopy />}
+                                    size="small"
+                                    className="absolute top-4 right-4"
+                                    onClick={() => {
+                                        void navigator.clipboard.writeText(status.backup_codes.join('\n') || '')
```
Comment (benjackwhite) [inline frontend/src/scenes/settings/user/TwoFactorAuthentication.tsx:88]: There is a `copyToClipboard` that does this with a fallback and som other niceties that is worth using

File: posthog/api/authentication.py (hunk 1 above)
Comment (benjackwhite) [reply posthog/api/authentication.py:261]: One thing I'm not sure about - I think normally using a backup code is a one time operation... Doing it should require them to re-setup 2fa right? 

Currently it doesn't look like the code is marked as used unless there is some under-the-hood thing I've missed? Appreciate this widens the scope but it feels potentially important 🤔

**Review by benjackwhite — CHANGES_REQUESTED** [review]

File: posthog/api/authentication.py (hunk 1 above)
Comment (zlwaterfield) [reply posthog/api/authentication.py:261]: Yeah, after going to bed I thought I should pull that change out into another PR and spend more time on it anyways. 
 I'm going to remove that change here and do it on another PR + look into your comment about resetting the 2FA.

File: frontend/src/scenes/authentication/login2FALogic.ts
Comment (zlwaterfield) [inline frontend/src/scenes/authentication/login2FALogic.ts:1]: This file is just 2FA clean up

**Review by benjackwhite — APPROVED** [review]

**Pull request merged**